storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+type RedisBackend struct {
+	host     string
+	password string
+	db       int
+	pool     *redis.Pool
+	// If set, path to a socket file overrides hostname
+	socketPath string
+	redsync    *redsync.Redsync
+	common.RedisConnector
+}
+
+type peerStore struct {
+	cfg Config
+	rb  *RedisBackend
+
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+var _ storage.PeerStore = &peerStore{}
+
+// populateProm aggregates metrics over all shards and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers uint64
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	for _, shard := range shards {
+		infohashes_list, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return
+		}
+		infohashes := infohashes_list.([]interface{})
+
+		InfohashLPrefix := shard + "_L_"
+		InfohashSPrefix := shard + "_S_"
+		InfohashPrefixLen := len(InfohashLPrefix)
+		InfohashesMap := make(map[string]bool)
+
+		for _, ih := range infohashes {
+			ih_str := string(ih.([]byte))
+			ih_str_infohash := ih_str[InfohashPrefixLen:]
+			if strings.HasPrefix(ih_str, InfohashLPrefix) {
+				numLeechers++
+				InfohashesMap[ih_str_infohash] = true
+			} else if strings.HasPrefix(ih_str, InfohashSPrefix) {
+				numSeeders++
+				InfohashesMap[ih_str_infohash] = true
+			} else {
+				log.Error("storage: invalid Redis state", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ih_str,
+				})
+			}
+		}
+
+		numInfohashes += uint64(len(InfohashesMap))
+	}
+
+	storage.PromInfohashesCount.Set(float64(numInfohashes))
+	storage.PromSeedersCount.Set(float64(numSeeders))
+	storage.PromLeechersCount.Set(float64(numLeechers))
+
+	log.Debug("storage: populateProm() aggregates metrics over all shards", log.Fields{
+		"numInfohashes": float64(numInfohashes),
+		"numSeeders":    float64(numSeeders),
+		"numLeechers":   float64(numLeechers),
+	})
+}
+
+// recordGCDuration records the duration of a GC sweep.
+func recordGCDuration(duration time.Duration) {
+	log.Debug("storage: recordGCDuration", log.Fields{"timeTaken(ms)": float64(duration.Nanoseconds()) / float64(time.Millisecond)})
+	storage.PromGCDurationMilliseconds.Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return timecache.NowUnixNano()
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: PutSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	// Update the peer in the swarm.
+	encoded_ih_S := IPver + "_S_" + ih.String()
```

Comment:
This should be `camelCase` at some point

Reply:
Done `camelCase` : )